HPCC-17484 Dynamic ESDL C++ Plugins

mayx opened the pull request:
- Add XSLTs to generate base service hpp and cpp files and CMakeLists.txt
  etc.
- Add option "cpp" to esdl command line to generate all files needed,using
  the XSLTs
- Add support for processing c++ plugin method type in esdl_binding

Signed-off-by: mayx <yanrui.ma@lexisnexisrisk.com>

<!-- Thank you for submitting a pull request to the HPCC project

 PLEASE READ the following before proceeding.

 This project only accepts pull requests related to open JIRA issues.
 If suggesting a new feature or change, please discuss it in a JIRA issue first.
 If fixing a bug, there should be an issue describing it with steps to reproduce.
 The title line of the pull request (and of each commit within it) should refer to the
 associated issue using the format:

 HPCC-nnnnn Short description of issue

 This will allow the Jira ticket to be automatically updated to refer to this pull request,
 and will ensure that the automatically-generated changelog is properly formatted.
 Where a pull request contains a single commit the pull request title will be set automatically,
 assuming that the commit has followed the proper guidelines.

 Please go over all the following points, and put an `x` in all the boxes that apply. You may find
 it easier to press the 'Create' button first then click on the checkboxes to edit the comment.
-->

## Type of change:
- [ ] This change is a bug fix (non-breaking change which fixes an issue).
- [x] This change is a new feature (non-breaking change which adds functionality).
- [ ] This change improves the code (refactor or other change that does not change the functionality)
- [ ] This change fixes warnings (the fix does not alter the functionality or the generated code)
- [ ] This change is a breaking change (fix or feature that will cause existing behavior to change).
- [ ] This change alters the query API (existing queries will have to be recompiled)

## Checklist:
- [x] My code follows the code style of this project.
  - [x] My code does not create any new warnings from compiler, build system, or lint.
- [x] The commit message is properly formatted and free of typos.
  - [x] The commit message title makes sense in a changelog, by itself.
  - [x] The commit is signed.
- [ ] My change requires a change to the documentation.
  - [ ] I have updated the documentation accordingly, or...
  - [ ] I have created a JIRA ticket to update the documentation.
  - [ ] Any new interfaces or exported functions are appropriately commented.
- [x] I have read the CONTRIBUTORS document.
- [x] The change has been fully tested:
  - [ ] I have added tests to cover my changes.
  - [x] All new and existing tests passed.
  - [ ] I have checked that this change does not introduce memory leaks.
  - [ ] I have used Valgrind or similar tools to check for potential issues.
- [x] I have given due consideration to all of the following potential concerns:
  - [x] Scalability
  - [x] Performance
  - [x] Security
  - [x] Thread-safety
  - [x] Premature optimization
  - [x] Existing deployed queries will not be broken
  - [x] This change fixes the problem, not just the symptom
  - [x] The target branch of this pull request is appropriate for such a change.
- [x] There are no similar instances of the same problem that should be addressed
  - [x] I have addressed them here
  - [ ] I have raised JIRA issues to address them separately
- [ ] This is a user interface / front-end modification
  - [ ] I have tested my changes in multiple modern browsers
  - [ ] The component(s) render as expected

## Testing:
<!-- Please describe how this change has been tested.-->

<!-- Thank you for taking the time to submit this pull request and to answer all of the above-->

mayx commented:
@afishbeck please review...

HPCCSmoketest commented:
Process of PR-12268, label: HPCC-17494-C++Plugins starts now.
The reason of this test is: Did not build yet.
Commit ID: 6f4c06806826a5de01f7a9a5722c9b97f6960b60
Estimated completion time is ~0.5 hour(s)

HPCCSmoketest commented:
Automated Smoketest: :white_check_mark:
OS: centos 7.4.1708 (Linux 3.10.0-327.28.3.el7.x86_64)
Sha: 6f4c06806826a5de01f7a9a5722c9b97f6960b60
Build: success
Build: success
Install HPCC Platform
HPCC Start: OK

Unit tests result:

| Test | total | passed | failed | errors | timeout | elaps | 
|---|---|---|---|---|---|---|
|unittest |104 |104 |0 |0 |0 |34 sec |
|wutoolTest(Dali) |19 |19 |0 |0 |0 |2 sec |
|wutoolTest(Cassandra) |19 |19 |0 |0 |0 |7 sec |



Regression test result:

| phase | total | pass | fail | elaps | 
|---|---|---|---|---|
|setup (hthor) |11 |11 |0 |24 sec  (00:00:24) |
|setup (thor) |11 |11 |0 |44 sec  (00:00:44) |
|setup (roxie) |11 |11 |0 |23 sec  (00:00:23) |
|test (hthor) |819 |819 |0 |257 sec  (00:04:17) |
|test (thor) |744 |744 |0 |597 sec  (00:09:57) |
|test (roxie) |893 |893 |0 |348 sec  (00:05:48) |


HPCC Stop: OK
Time stats:

| Prep time | Build time | Package time | Install time | Start time | Test time | Stop time | Summary | 
|---|---|---|---|---|---|---|---|
|   31 sec (00:00:31) |  194 sec (00:03:14) |    0 sec (00:00:00) |    2 sec (00:00:02) |   18 sec (00:00:18) | 1488 sec (00:24:48) |   21 sec (00:00:21) | 1754 sec (00:29:14) |

mayx committed:
HPCC-17484 Dynamic ESDL C++ Plugins

- Add XSLTs to generate base service hpp and cpp files and CMakeLists.txt
  etc.
- Add option "cpp" to esdl command line to generate all files needed,using
  the XSLTs
- Add support for processing c++ plugin method type in esdl_binding

Signed-off-by: mayx <yanrui.ma@lexisnexisrisk.com>

HPCCSmoketest commented:
Process of PR-12268, label: HPCC-17494-C++Plugins starts now.
The reason of this test is: New commit.
Commit ID: 7d148a463711cac84c5a0c32b6aa8869419fb4c1
Estimated completion time is ~0.5 hour(s)

HPCCSmoketest commented:
Automated Smoketest: :white_check_mark:
OS: centos 7.4.1708 (Linux 3.10.0-327.28.3.el7.x86_64)
Sha: 7d148a463711cac84c5a0c32b6aa8869419fb4c1
Build: success
Build: success
Install HPCC Platform
HPCC Start: OK

Unit tests result:

| Test | total | passed | failed | errors | timeout | elaps | 
|---|---|---|---|---|---|---|
|unittest |104 |104 |0 |0 |0 |32 sec |
|wutoolTest(Dali) |19 |19 |0 |0 |0 |2 sec |
|wutoolTest(Cassandra) |19 |19 |0 |0 |0 |7 sec |



Regression test result:

| phase | total | pass | fail | elaps | 
|---|---|---|---|---|
|setup (hthor) |11 |11 |0 |26 sec  (00:00:26) |
|setup (thor) |11 |11 |0 |47 sec  (00:00:47) |
|setup (roxie) |11 |11 |0 |21 sec  (00:00:21) |
|test (hthor) |819 |819 |0 |249 sec  (00:04:09) |
|test (thor) |744 |744 |0 |585 sec  (00:09:45) |
|test (roxie) |893 |893 |0 |352 sec  (00:05:52) |


HPCC Stop: OK
Time stats:

| Prep time | Build time | Package time | Install time | Start time | Test time | Stop time | Summary | 
|---|---|---|---|---|---|---|---|
|   22 sec (00:00:22) |  198 sec (00:03:18) |    0 sec (00:00:00) |    2 sec (00:00:02) |   18 sec (00:00:18) | 1475 sec (00:24:35) |   21 sec (00:00:21) | 1736 sec (00:28:56) |

HPCCSmoketest commented:
Process of PR-12268, label: HPCC-17494-C++Plugins starts now.
The reason of this test is: New commit.
Commit ID: a528e3f74c8bd890a7594d4e6923707ed3e89303
Estimated completion time is ~0.5 hour(s)

HPCCSmoketest commented:
Automated Smoketest: :white_check_mark:
OS: centos 7.4.1708 (Linux 3.10.0-327.28.3.el7.x86_64)
Sha: a528e3f74c8bd890a7594d4e6923707ed3e89303
Build: success
Build: success
Install HPCC Platform
HPCC Start: OK

Unit tests result:

| Test | total | passed | failed | errors | timeout | elaps | 
|---|---|---|---|---|---|---|
|unittest |104 |104 |0 |0 |0 |34 sec |
|wutoolTest(Dali) |19 |19 |0 |0 |0 |2 sec |
|wutoolTest(Cassandra) |19 |19 |0 |0 |0 |7 sec |



Regression test result:

| phase | total | pass | fail | elaps | 
|---|---|---|---|---|
|setup (hthor) |11 |11 |0 |25 sec  (00:00:25) |
|setup (thor) |11 |11 |0 |45 sec  (00:00:45) |
|setup (roxie) |11 |11 |0 |21 sec  (00:00:21) |
|test (hthor) |819 |819 |0 |248 sec  (00:04:08) |
|test (thor) |744 |744 |0 |597 sec  (00:09:57) |
|test (roxie) |893 |893 |0 |348 sec  (00:05:48) |


HPCC Stop: OK
Time stats:

| Prep time | Build time | Package time | Install time | Start time | Test time | Stop time | Summary | 
|---|---|---|---|---|---|---|---|
|   31 sec (00:00:31) |  195 sec (00:03:15) |    0 sec (00:00:00) |    2 sec (00:00:02) |   18 sec (00:00:18) | 1502 sec (00:25:02) |   19 sec (00:00:19) | 1767 sec (00:29:27) |

afishbeck commented:
@timothyklemm can you take a first pass at reviewing this?

mayx commented on esp/services/esdl_svc_engine/esdl_binding.cpp:
```diff
@@ -449,6 +491,20 @@ void EsdlServiceImpl::configureTargets(IPropertyTree *cfg, const char *service)
+                    DBGLOG("C++ plugin name is missing for method %s", method);
+                else
+                {
+                    IEspPlugin* plugin = localCppPluginMap.getValue(pluginName);
```
The local map is to keep track of which library has been reloaded in the current pass. When there're multiple entry points in the library, we only need to reload the library once. The loop is for each method/entry point, so reloading the library for every method is a waste.

timothyklemm replied:
My bad. I didn't make the connection with the condition in configureCppMethod.

mayx commented on esp/services/esdl_svc_engine/esdl_binding.cpp:
```diff
@@ -483,6 +540,10 @@ inline EsdlMethodImplType getEsdlMethodImplType(const char *querytype)
             return EsdlMethodImplProxy;
         if (strieq(querytype, "java"))
             return EsdlMethodImplJava;
+        if (strieq(querytype, "java"))
```
Looks like, will remove.

mayx commented on esp/services/esdl_svc_engine/esdl_binding.cpp:
```diff
@@ -651,6 +712,44 @@ void EsdlServiceImpl::handleServiceRequest(IEspContext &context,
+            ctxbuf.append("</EsdlContext>");
+
+            context.addTraceSummaryTimeStamp(LogNormal, "srt-cppcall");
+            xproc(ctxbuf.str(), reqcontent.str(), origResp);
```
This release the plugin is to accomplish the same as the Java plugin. We'll add more info to the context. Also we have to rethink configuration and logging for both Java and C++. Let's create new Jira issues for those.

mayx commented on esp/xslt/esdl2cpp_cmake.xslt:
```diff
@@ -0,0 +1,49 @@
+
+if (("${HPCC_SOURCE_DIR}" STREQUAL "") OR ("${HPCC_BUILD_DIR}" STREQUAL "") OR ("${CMAKE_BUILD_TYPE}" STREQUAL
+ ""))
+    message (FATAL_ERROR "Please specify HPCC_SOURCE_DIR, HPCC_BIULD_DIR and CMAKE_BUILD_TYPE")
```
Will fix typo.

mayx commented on esp/xslt/esdl2cpp_srvbasecpp.xslt:
```diff
@@ -0,0 +1,357 @@
+    <xsl:for-each select="EsdlMethod">
+    <xsl:text>int </xsl:text><xsl:value-of select="$servicebase"/>::<xsl:value-of select="@name"/><xsl:text>(const char* CtxStr, const char* ReqStr, StringBuffer&amp; RespStr)
+{</xsl:text>
+    DBGLOG("Received request: %s", ReqStr);
```
Will remove this line.

mayx commented on esp/xslt/esdl2cpp_srvbasehpp.xslt:
```diff
@@ -0,0 +1,273 @@
+</xsl:text>
+    </xsl:template>
+
+    <xsl:template name="outputCppPrimitive">
```
I tried it and found it can't be easily done. The transform seems to require the full path of the included file, relative path doesn't work. But the full path isn't known until after installation. We could do something fancy like generating the XSLT file at installation time or run time, but I don't think it's worth the effort.

HPCCSmoketest commented:
This PR (12268) is in the 1/1 position of the queue. Estimated start time is after ~0.5 hour(s)

mayx commented on tools/esdlcmd/esdlcmd_core.cpp:
```diff
@@ -801,6 +801,313 @@ class Esdl2JavaCmd : public EsdlHelperConvertCmd
+        filefullpath.clear().append(sourcedir).append(PATHSEPCHAR).append("primitivetypes.hpp");
+        if (!checkFileExists(filefullpath.str()))
+        {
+            VStringBuffer primitivetypesFile("%s%cexamples%cEsdlExample%cCppPlugin%cprimitivetypes.hpp", INSTALL_DIR, PATHSEPCHAR,
```
Agree, will change to generating the file.

mayx commented:
@timothyklemm replied to your comments, uploaded changes. Please review again...

HPCCSmoketest commented:
Process of PR-12268, label: HPCC-17494-C++Plugins starts now.
The reason of this test is: New commit.
Commit ID: b9e6b414ad2fe26bdd806d67a3b54a5b4b1caf98
Estimated completion time is ~0.5 hour(s)

HPCCSmoketest commented:
Automated Smoketest: :x:
OS: centos 7.4.1708 (Linux 3.10.0-327.28.3.el7.x86_64)
Sha: b9e6b414ad2fe26bdd806d67a3b54a5b4b1caf98
Build: success
Build: success
Install HPCC Platform
HPCC Start: Fail
HPCC start:
/mnt/disk1/home/vamosax/smoketest/PR-12268/hpcc/opt/HPCCSystems/sbin/configgen: error while loading shared libraries: libeclrtl.so: cannot open shared object file: No such file or directory
There are no components configured to run on this node ...
 
HPCC status:
/mnt/disk1/home/vamosax/smoketest/PR-12268/hpcc/opt/HPCCSystems/sbin/configgen: error while loading shared libraries: libeclrtl.so: cannot open shared object file: No such file or directory
There are no components configured to run on this node ...
res: 
Archive HPCC logs into /mnt/disk1/home/vamosax/smoketest/PR-12268/HPCCSystems-logs-2019-03-13_19-16-58.
Check core files.
There is no core file in /var/lib/HPCCSystems/
Local 0 running components
Local HPCC System already stopped.
Remove Cassandra leftovers.
Finish at: 2019-03-13_19-20-37
time logs started
Prep time   :    26 sec (00:00:26)
Build time  :   190 sec (00:03:10)
Package time:     0 sec (00:00:00)
Install time:     1 sec (00:00:01)
Start time  :     0 sec (00:00:00)
Test time   :     0 sec (00:00:00)
Stop time   :     0 sec (00:00:00)
Summary     :   217 sec (00:03:37)
time logs end
All done.

[Build error:](https://gist.github.com/492fbf575ac42dfc72b327bac662756e)
Build log [Build log](https://gist.github.com/HPCCSmoketest/492fbf575ac42dfc72b327bac662756e#file-result-19-03-13-19-16-39-log)

timothyklemm commented on esp/xslt/CMakeLists.txt:
```diff
@@ -45,6 +45,12 @@ FOREACH( iFILES
+    ${CMAKE_CURRENT_SOURCE_DIR}/esdl2cpp_srvhpp.xslt
+    ${CMAKE_CURRENT_SOURCE_DIR}/esdl2cpp_srvcpp.xslt
+    ${CMAKE_CURRENT_SOURCE_DIR}/esdl2cpp_cmake.xslt
+    ${CMAKE_CURRENT_SOURCE_DIR}/esdl2cpp_types.xslt
```
Did this file get added? I only see the example copy of primitivetypes.hpp.

mayx replied:
Good catch. Now the file esdl2cpp_types.xslt is added, and the hpp file is removed.

mayx commented:
@timothyklemm change pushed, please review again...

HPCCSmoketest commented:
Process of PR-12268, label: HPCC-17494-C++Plugins starts now.
The reason of this test is: New commit.
Commit ID: 8a427e70b24c2b4be26efc1e249eeaf30c1f212a
Estimated completion time is ~0.5 hour(s)

HPCCSmoketest commented:
Automated Smoketest: :white_check_mark:
OS: centos 7.4.1708 (Linux 3.10.0-327.28.3.el7.x86_64)
Sha: 8a427e70b24c2b4be26efc1e249eeaf30c1f212a
Build: success
Build: success
Install HPCC Platform
HPCC Start: OK

Unit tests result:

| Test | total | passed | failed | errors | timeout | elaps | 
|---|---|---|---|---|---|---|
|unittest |104 |104 |0 |0 |0 |35 sec |
|wutoolTest(Dali) |19 |19 |0 |0 |0 |2 sec |
|wutoolTest(Cassandra) |19 |19 |0 |0 |0 |7 sec |



Regression test result:

| phase | total | pass | fail | elaps | 
|---|---|---|---|---|
|setup (hthor) |11 |11 |0 |26 sec  (00:00:26) |
|setup (thor) |11 |11 |0 |44 sec  (00:00:44) |
|setup (roxie) |11 |11 |0 |21 sec  (00:00:21) |
|test (hthor) |819 |819 |0 |262 sec  (00:04:22) |
|test (thor) |744 |744 |0 |606 sec  (00:10:06) |
|test (roxie) |893 |893 |0 |349 sec  (00:05:49) |


HPCC Stop: OK
Time stats:

| Prep time | Build time | Package time | Install time | Start time | Test time | Stop time | Summary | 
|---|---|---|---|---|---|---|---|
|   38 sec (00:00:38) |  195 sec (00:03:15) |    0 sec (00:00:00) |    2 sec (00:00:02) |   18 sec (00:00:18) | 1504 sec (00:25:04) |   20 sec (00:00:20) | 1777 sec (00:29:37) |
